url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+
+                // Wake up the eval to take a step for us.
+                this.notifyAll();
+                // We put ourselves in wait mode until the eval has taken a step.
+                // The eval has to get us out of the wait queue.
+                this.wait();
+                stepcount ++;
+            }
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+        }
+    }
+
+    public synchronized Eval2.EvalStack getStack()
+    {
+        // No guard because you could always get the last stack for examination,
+        // even post mortem.
+
+        try
+        {
+            // We cant access the stack while the
+            // eval is working on it. So we are patient.
+            while(!halted) this.wait();
+            // If the eval is halted, we can access its stack.
+            return stack;
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+            return null;
+        }
+    }
+
+    /**
+     * Obtain the last result produced by the eval under trace.
+     * Note that the result can get overwritten by subsequent evaluations, especially by
+     * commands using the eval.
+     *
+     * @return
+     */
+    public synchronized Object getResult()
+    {
+        // No guard because you always could get the last result for examination,
+        // even post mortem.
+
+        try
+        {
+            while(!halted) this.wait();
+            return result;
+        }
+        catch (InterruptedException e)
+        {
+            // The wait could be interrupted because the
+            // eval thread ended.
+            halted = true;
+            return result;
+        }
+    }
+
+    /**
+     * Check if the eval under trace has already reached a result.
+     * Note that the result could also be null, it is a valid result.
+     *
+     * @return
+     */
+    public synchronized boolean hasResult()
+    {
+        // No guard because you always could get the last result for examination,
+        // even post mortem.
+
+        try
+        {
+            while(!halted) this.wait();
+            return hasResult;
+        }
+        catch (InterruptedException e)
+        {
+            // The wait could be interrupted because the
+            // eval thread ended.
+            halted = true;
+            return hasResult;
+        }
+    }
+
+    /**
+     * Stop tracing. No operation is possible after the tracer has halted.
+     * All resources held by the eval being traced will be released.
+     * Stopping a tracer is necessary in order not to clutter the memory.
+     *
+     */
+   public synchronized void terminate()
+   {
+       if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+       try
+        {
+            while(!halted) this.wait();
+            worker.interrupt();
+            worker = null;
+            halted = true;
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)